### PR TITLE
engine: support setting memory limit for raft engine

### LINKED ISSUE #12255: two tikv oom after inject tikv network-loss and recovery for some time
### What version of TiKV are you using?
[2022/03/24 04:13:50.851 +08:00] [INFO] [client.go:376] ["Cluster version information"] [type=pd] [version=6.1.0-nightly] [git_hash=1ac0ad691260dabb61a25f30359e996a968ed857]
[2022/03/24 04:13:50.851 +08:00] [INFO] [client.go:376] ["Cluster version information"] [type=tikv] [version=6.0.0-alpha] [git_hash=869b953e798cabf29872fd17d526a7061437aec2]
[2022/03/24 04:13:50.851 +08:00] [INFO] [client.go:376] ["Cluster version information"] [type=tidb] [version=6.1.0-nightly] [git_hash=b9bacad6dafabf5e2dfafee8e50ac66785e911b6]

### What operating system and CPU are you using?
8core、16GB
2tidb、3pd、5tikv（5replicas）
### Steps to reproduce
https://tcms.pingcap.net/dashboard/executions/plan/662849
test data：
 {{[tpcc] []} {s3://benchmark/tpcc10000 tpcc10000 10000 64 2013,1213,1105,1205,8022,8027,8028,9004,9007,1062} {s3://benchmark/sysbench_64_7000w sysbench_64_7000w 64 70000000 64 2013,1213,1105,1205,8022,8027,8028,9004,9007,1062} {0} {[]} {false       }} 

1、[2022/03/24 04:13:51.083 +08:00] [INFO] [cmd.go:124] ["Start remote command"] [cmd="go-tpc tpcc run -D tpcc10000 --host tc-tidb.endless-oltp-tps-662849-1-968 -P4000 --warehouses 10000 -T 64 --time 36000m --ignore-error '2013,1213,1105,1205,8022,8027,8028,9004,9007,1062'"] [nodename=benchtoolset]
2、inject fault
[2022/03/24 04:24:51.173 +08:00] [INFO] [chaos.go:86] ["Run chaos"] [name=network-loss] [selectors="[endless-oltp-tps-662849-1-968/tc-tikv-1]"] [experiment="{\"Duration\":\"\",\"Scheduler\":null,\"Loss\":\"84\",\"Correlation\":\"25\"}"]
[2022/03/24 04:24:51.175 +08:00] [INFO] [chaos.go:86] ["Run chaos"] [name=network-loss] [selectors="[endless-oltp-tps-662849-1-968/tc-tikv-0]"] [experiment="{\"Duration\":\"\",\"Scheduler\":null,\"Loss\":\"84\",\"Correlation\":\"25\"}"]
3、recovery fault
[2022/03/24 05:06:51.203 +08:00] [INFO] [chaos.go:151] ["Clean chaos"] [name=network-loss] [chaosId="ns=endless-oltp-tps-662849-1-968,kind=network-loss,name=network-loss-pdhgfxcy,spec=&k8s.ChaosIdentifier{Namespace:\"endless-oltp-tps-662849-1-968\", Name:\"network-loss-pdhgfxcy\", Spec:NetworkLossSpec{Duration: \"\", Scheduler: <nil>, Loss: \"84\", Correlation: \"25\"}}"]
[2022/03/24 05:06:51.203 +08:00] [INFO] [chaos.go:151] ["Clean chaos"] [name=network-loss] [chaosId="ns=endless-oltp-tps-662849-1-968,kind=network-loss,name=network-loss-zfevalyq,spec=&k8s.ChaosIdentifier{Namespace:\"endless-oltp-tps-662849-1-968\", Name:\"network-loss-zfevalyq\", Spec:NetworkLossSpec{Duration: \"\", Scheduler: <nil>, Loss: \"84\", Correlation: \"25\"}}"]

### What did you expect?
all tikv are normal

### What did happened?
two tikv oom at 2022/03/24 06:02 and 06:26
tikv0 memory start to rise form 2022/03/24 05:10 and oom at 06:26
tikv1 memory start to rise form 2022/03/24 05:08 and oom at 06:00
image
image

image
image
image
image
image
image
image
image

### PR DESCRIPTION
**Pull request by tabokie**
### What is changed and how it works?

Issue Number: Close #12255

What's Changed:

```commit-message
Support setting memory limit for raft engine
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests
- Unit test
- Integration test

Side effects

- Performance regression when follower down and memory not enough

### Release note 

```release-note
Support setting memory limit for raft engine
```

### CODE REVIEW

File: components/raft_log_engine/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,7 @@ num_cpus = "1"
 online_config = { path = "../online_config" }
 protobuf = "2"
 raft = { version = "0.7.0", default-features = false, features = ["protobuf-codec"] }
-raft-engine = "0.1.0"
+raft-engine = { git = "https://github.com/tikv/raft-engine.git", features = ["swap"] }
```
[the remark this answers is not in the document]
Comment (tabokie) [reply components/raft_log_engine/Cargo.toml:17]: Discussed in https://github.com/tikv/tikv/pull/11766#discussion_r848211848, will bump version after QA approves the final version.

**Review by Connor1996 — APPROVED** [review]
LGTM

File: tests/integrations/config/mod.rs [hunk 2]
```diff
@@ -16,7 +16,7 @@ use engine_traits::PerfLevel;
 use file_system::{IOPriority, IORateLimitMode};
 use kvproto::encryptionpb::EncryptionMethod;
 use pd_client::Config as PdConfig;
-use raft_log_engine::RecoveryMode;
+use raft_log_engine::{ReadableSize as RaftEngineReadableSize, RecoveryMode};
```
Comment (Connor1996) [inline tests/integrations/config/mod.rs:19]: why is `ReadableSize` in raft_log_engine

Comment (tabokie) [reply tests/integrations/config/mod.rs:19]: Config is defined in Raft Engine, tikv's ReadableSize is copied there.

**Pull request merged**